backend/status/status_source.go

```diff
@@ -0,0 +1,222 @@
+package status
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net"
+	"net/http"
+	"os"
+	"syscall"
+
+	grpccodes "google.golang.org/grpc/codes"
+	grpcstatus "google.golang.org/grpc/status"
+)
+
+// Source type defines the status source.
+type Source string
+
+const (
+	// SourcePlugin status originates from plugin.
+	SourcePlugin Source = "plugin"
+
+	// SourceDownstream status originates from downstream service.
+	SourceDownstream Source = "downstream"
+
+	// DefaultSource is the default [Source] that should be used when it is not explicitly set.
+	DefaultSource Source = SourcePlugin
+)
+
+// IsValid return true if es is [SourceDownstream] or [SourcePlugin].
+func (s Source) IsValid() bool {
+	return s == SourceDownstream || s == SourcePlugin
+}
+
+// String returns the string representation of s. If s is not valid, [DefaultSource] is returned.
+func (s Source) String() string {
+	if !s.IsValid() {
+		return string(DefaultSource)
+	}
+
+	return string(s)
+}
+
+// SourceFromHTTPError returns a [Source] based on provided error.
+func SourceFromHTTPError(err error) Source {
+	if IsDownstreamHTTPError(err) {
+		return SourceDownstream
+	}
+	return SourcePlugin
+}
+
+// ErrorSourceFromStatus returns a [Source] based on provided HTTP status code.
+func SourceFromHTTPStatus(statusCode int) Source {
+	switch statusCode {
+	case http.StatusMethodNotAllowed,
+		http.StatusNotAcceptable,
+		http.StatusPreconditionFailed,
+		http.StatusRequestEntityTooLarge,
+		http.StatusRequestHeaderFieldsTooLarge,
+		http.StatusRequestURITooLong,
+		http.StatusExpectationFailed,
+		http.StatusUpgradeRequired,
+		http.StatusRequestedRangeNotSatisfiable,
+		http.StatusNotImplemented:
+		return SourcePlugin
+	}
+
+	return SourceDownstream
+}
+
+type errorWithSourceImpl struct {
+	source Source
+	err    error
+}
+
+// DownstreamError creates a new error with status [SourceDownstream].
+func DownstreamError(err error) error {
```

Comment:
I looked into this further and I think we should avoid overriding the status if it already exists. I reviewed the usage of [errorsource.DownstreamError() which lets you decide whether or not to override the source](https://github.com/search?q=org%3Agrafana%20.DownstreamError(&type=code), and in all cases, plugin developers choose not to override it. Since we are introducing this new method that might be used frequently, I'd suggest not overriding the status if it already exists. Instead, we could create a method like `DownstreamErrorWithOverride` for the rare cases where users want to override the status.

Reply:
First, this is not really a new function. And for now the experimental errorsource package are mostly used, but planned to be removed soon.

The risk of not overriding is that we drop important errors/context/information. Further, if error already contains a downstream source and you call `DownstreamError` with an error with a plugin source and not override we might drop what is actually a real plugin error/problem. Not sure what the approach should be here, but I would be in favor of the current way just overriding. In the opposite way I guess you want to override though 😱  

So maybe the logic needs to 
- check if incoming error has plugin source it will override a downstream source 
- check if incoming error has downstream source and it will not override a plugin source, but we might be able to unwrap the error to not lose the details or something.
 
I would rather punt on this for now until we've agreed on the details how this should work.